kvserver/rangefeed: move drainAllocation call

stevendanna opened the pull request:
Previously, drainAllocations was called by registry.Unregister. While this was nice as it was a single call site, it seems a bit off to me because it ties to lifetime of the allocations in the registration's buffer to the registry rather than to the registration's output loop which is the reader of that buffer.

Epic: none
Release note: None

stevendanna committed:
kvserver/rangefeed: move drainAllocation call

Previously, drainAllocations was called by registry.Unregister. While
this was nice as it was a single call site, it seems a bit off to me
because it ties to lifetime of the allocations in the registration's
buffer to the registry rather than to the registration's output loop
which is the reader of that buffer.

Epic: none
Release note: None

wenyihu6 commented:
Apologies for my delayed responses on the other PR discussion, still catching up on the threads. I thought we would want to call `drainAllocations` only after deleting the registration from the registry tree but not right after `runOutputLoop`. Even after `runOutputLoop` ends and we call `disconnect` to invoke cleanup async, updates may still be sent to the registration channel from the processor until the registration is removed from the tree. Does that seem right to you?

wenyihu6 reviewed (commented):
Apologies for my delayed responses on the other PR discussion, still catching up on the threads. I thought we would want to call `drainAllocations` only after deleting the registration from the registry tree but not right after `runOutputLoop`. Even after `runOutputLoop` ends and we call `disconnect` to invoke cleanup async, updates may still be sent to the registration channel from the processor until the registration is removed from the tree. Does that seem right to you?

stevendanna commented:
@wenyihu6 I _think_ we can solve that by not accepting events if we are disconnected?

stevendanna committed:
kvserver/rangefeed: don't add events to buffer after disconnect

Once disconnected, it is likely that nothing is reading this message
from the buffer. Adding it to the buffer is thus unnecessarily and
complicates when we are allowed to safely drain the buffer.

Epic: none
Release note: None

wenyihu6 commented:
Right, I agree that we could solve this by checking for `r.mu.disconnected` within `r.publish`. I like the current structure of buffered registration as it provides a natural spot for drainAllocations, but since we plan to drain catch up buffer right after runOutputLoop for unbuffered registrations, making this change would give us more parallel with our upcoming PR. wdyt?

stevendanna commented:
> I like the current structure of buffered registration as it provides a natural spot for drainAllocations

I go back and forth on this. It is natrual in the sense that the registration knows it won't be sending more events so that is a clear place to call it. But it seems un-natural in that some other object (the registry) needs to know details about the registration.

> making this change would give us more parallel with our upcoming PR. wdyt?

Yeah, this is what I'm thinking as well. I've pushed a commit to check disconnected in publish. Let's see what CI says.

stevendanna commented:
CI passes but I think perhaps a new test is in order here.  I can remove these drain allocation calls and nothing fails. I think there might be a test we added on the larger branch I can pull in somewhere.

tbg commented on pkg/kv/kvserver/rangefeed/buffered_registration.go:
```diff
@@ -120,6 +120,10 @@ func (br *bufferedRegistration) publish(
 	if br.mu.overflowed {
```
drive-by: is it intentional that publish holds an exclusive mutex below even in the test case in which we do a blocking send?

wenyihu6 replied:
Hmm, I'm not entirely sure if I follow. I think we need to hold an exclusive mutex here since we write to this variable here https://github.com/cockroachdb/cockroach/blob/5983b096046893b1a95dcfe9fbf1150d4fb95201/pkg/kv/kvserver/rangefeed/buffered_registration.go#L146 and then read it in another goroutine, `runOutputLoop` https://github.com/cockroachdb/cockroach/blob/5983b096046893b1a95dcfe9fbf1150d4fb95201/pkg/kv/kvserver/rangefeed/buffered_registration.go#L216. Perhaps Steven might have some insights.


Which test case are you referring to? Are you talking about the one where we cause `stream.Send` to be blocked https://github.com/cockroachdb/cockroach/blob/c48acff77f042cc57470031db7191594f1621f64/pkg/kv/kvserver/rangefeed/unbuffered_sender_test.go#L167? Or are you referring more generally to a send to the gRPC stream without buffering events?
<!-- Reviewable comment -OC9PkLEAXaW34cNazyV:b8nu196 -->
<!-- Sent from Reviewable.io -->

tbg commented on pkg/kv/kvserver/rangefeed/buffered_registration.go:
```diff
@@ -120,6 +120,10 @@ func (br *bufferedRegistration) publish(
 	if br.mu.overflowed {
 		return
 	}
+	if br.mu.disconnected {
```
in the commit message, what do you men by "adding it to the buffer is only more work later"? What exactly happens to an event in a disconnected buffer?

wenyihu6 replied:
> What exactly happens to an event in a disconnected buffer?

After `registration.disconnect()` is called, additional events may still be buffered from the processor https://github.com/cockroachdb/cockroach/blob/a4cbf7d7ec37fbb351e453bf2c7314bac9cbfe9a/pkg/kv/kvserver/rangefeed/buffered_registration.go#L125. The buffer is not drained until `reg.Unregister` is called which officially unsubscribe the registration from the processor and triggers the draining process https://github.com/cockroachdb/cockroach/blob/a4cbf7d7ec37fbb351e453bf2c7314bac9cbfe9a/pkg/kv/kvserver/rangefeed/registry.go#L334-L341. `reg.Unregister` currently happens in the separate per-registration goroutine after `runOutputLoop` returns here https://github.com/cockroachdb/cockroach/blob/3bf870a87e8dd943fe9407dce277e20ae1718d6c/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L351-L357. They should all occur within a short time frame.

In this PR, we changed the flow a bit. After `registration.disconnect()` is called, no additional events can be buffered. Buffer draining now happens within the defer statement of `runOutputLoop`, which may occur before the processor's official unregistration (`reg.Unregister`).


> in the commit message, what do you men by "adding it to the buffer is only more work later"?

I believe this is what Steven meant by `adding it to the buffer only creates more work later.  For events buffered after `r.Disconnect()` but before `p.reg.Unregister()` happens, they are unlikely to be published, as `runOutputLoop` would be in the process of being torn down. Therefore, these events will just be buffered only to be drained later during `p.reg.Unregister()`. Instead, we can simply avoid buffering these events during by checking for r.disconnected.

Aside - currently, `p.reg.Unregister()` runs in a separate per-registration goroutine, but we plan to make it async and delegate the task to the processor goroutine here https://github.com/cockroachdb/cockroach/blob/9a1fe4e3f8758a6cc5c8fb47273dfd3d7865fd51/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L167. Moving buffer draining to the end of separate goroutine `runOutputLoop` would also help reduce processor's work.
<!-- Reviewable comment -OC9SlOI91S9c0crqTj7:bb5ktpe -->
<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134820#-:-OC9OpF_7LSE5MWmJYx7:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @stevendanna and @tbg)

___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134820#-OC9OsCoCG9W7e7QVLrC:-OC9OsCoCG9W7e7QVLrD:bsays83) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4278cfc09cc72225ef12939937516ddf10d27c1/pkg/kv/kvserver/rangefeed/buffered_registration.go#L125)):*
> ```Go
> 	if br.mu.disconnected {
> 		return
> 	}
> ```

minor nit: combine this into `if br.mu.overflowed || br.mu.disconnected`
___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 228 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134820#-OC9NmYiDVxDYVvqtU3f:-OC9NmYiDVxDYVvqtU3g:bf32xmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4278cfc09cc72225ef12939937516ddf10d27c1/pkg/kv/kvserver/rangefeed/buffered_registration.go#L228)):*
> ```Go
> 
> func (br *bufferedRegistration) runOutputLoop(ctx context.Context, _forStacks roachpb.RangeID) {
> 	defer br.drainAllocations(ctx)
> ```

Could we add a sentence here to clarify that no additional events should be buffered, either concurrently or after this method returns?
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134820#-OC9N8oN2Xov72i3k074:-OC9N8oN2Xov72i3k075:boe2v94) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4278cfc09cc72225ef12939937516ddf10d27c1/pkg/kv/kvserver/rangefeed/registry.go#L331)):*
> ```Go
> // registration has already been disconnected, this is intended only to clean
> // up the registry.
> // We also drain all pending events for the sake of memory accounting. To do
> ```

This comment is outdated now.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134820)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @stevendanna and @tbg)

___
*[`pkg/kv/kvserver/rangefeed/buffered_registration.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134820#-OC9OsCoCG9W7e7QVLrC:-OCAkRW795V4vCSzFT3I:br8fwgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4278cfc09cc72225ef12939937516ddf10d27c1/pkg/kv/kvserver/rangefeed/buffered_registration.go#L125)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

minor nit: combine this into `if br.mu.overflowed || br.mu.disconnected`
</blockquote></details>

btw feel free to ignore this comment - just a preference, your call


<!-- Sent from Reviewable.io -->

stevendanna commented:
The extended CI failure is unrelated: https://github.com/cockroachdb/cockroach/pull/136111

stevendanna commented:
bors r+
